raiden/tests/integration/test_matrix_transport.py

```diff
@@ -305,3 +308,84 @@ def on_message(self, _, message):
 
     transport0.stop()
     transport1.stop()
+
+
+def test_matrix_message_retry(
+    skip_if_not_matrix,
+    local_matrix_server,
+    private_rooms,
+    retry_interval,
+    retries_before_backoff,
+):
+    """ Test the retry mechanism implemented into the matrix client.
+    The test creates a transport and sends a message. Given that the
+    receiver was online, the initial message is sent but the receiver
+    doesn't respond in time and goes offline. The retrier should then
+    wait for the `retry_interval` duration to pass and send the message
+    again but this won't work because the receiver is offline. Once
+    the receiver comes back again, the message should be sent again.
+    """
+    partner_address = make_address()
+
+    transport = MatrixTransport({
+        'discovery_room': 'discovery',
+        'retries_before_backoff': retries_before_backoff,
+        'retry_interval': retry_interval,
+        'server': local_matrix_server,
+        'server_name': 'matrix.local.raiden',
+        'available_servers': [],
+        'private_rooms': private_rooms,
+    })
+    transport._send_raw = MagicMock()
+    raiden_service = MockRaidenService(None)
+
+    transport.start(
+        raiden_service,
+        raiden_service.message_handler,
+        None,
+    )
+    transport.log = MagicMock()
+
+    # Receiver is online
+    transport._address_to_presence[partner_address] = UserPresence.ONLINE
+
+    queueid = QueueIdentifier(
+        recipient=partner_address,
+        channel_identifier=CHANNEL_IDENTIFIER_GLOBAL_QUEUE,
+    )
+    chain_state = raiden_service.wal.state_manager.current_state
+
+    retry_queue = _RetryQueue(transport, partner_address)
+    retry_queue.start()
+
+    # Send the initial message
+    message = Processed(0)
+    message.sign(transport._raiden_service.private_key)
+    chain_state.queueids_to_queues[queueid] = [message]
+    retry_queue.enqueue_global(message)
+
+    gevent.sleep(1)
+
+    transport._send_raw.call_count = 1
+
+    # Receiver goes offline
+    transport._address_to_presence[partner_address] = UserPresence.OFFLINE
+
+    gevent.sleep(retry_interval)
+
+    transport.log.debug.assert_called_with(
+        'Partner not reachable. Skipping.',
+        partner=pex(partner_address),
+        status=UserPresence.OFFLINE,
+    )
+
+    # Retrier did not call send_raw given that the receiver is still offline
+    assert transport._send_raw.call_count == 1
+
+    # Receiver goes offline
+    transport._address_to_presence[partner_address] = UserPresence.ONLINE
+
+    gevent.sleep(retry_interval)
+
+    # Retrier now should have sent the message again
+    assert transport._send_raw.call_count == 2
```

Comment:
Can you somehow make sure that he receives the messages because of the retry mechanism and not because of coming online and syncing with the server and pulling all old messages thanks to the change done [here](https://github.com/raiden-network/raiden/pull/3090)?

Reply:
> The sync here doesn't kickin because we never stop the transport.

Ah you just change their `_address_to_presence` to make them appear offline for the partner, right? So I think it's indeed no problem.

> The retrier is what calls _send_raw to send a message.

Sure, but if you had the partner go offline and then restart the sync would have kicked in, no?

Reply:
The test deals with the sender side of the exchange... so to answer your question, the change in `_address_to_presence` is indeed to change the view for the sender about the status of the receiver. But even when we set the partner's state to online, the sync won't kick in from our perspective a sender.

Reply:
Ahhh right. I get it now. It's all completely from the sender's perspective. Thanks.